File: Vol.10(1900-1999)/#1916 Time Zones/p1916.cpp
// 1886441 2009-05-28 03:57:44 Accepted  1916 C++ 0 184 呆滞的慢板 

// 时区转换，就是字符串形式的时间和分钟数之间的转换。
// 实际时间表示上，如果对 am-pm 的时间表示方式不是太熟悉的话， 
// 搞起来有点恶心。时区的输入处理稍要技巧，手工打一个表然后 map 就差不多了。 

#include <iostream>
#include <string>
#include <map>
using namespace std;

map<string, int> MP;

string name[32] = {
    "UTC",  "GMT",  "BST",  "IST",  "WET",
    "WEST", "CET",  "CEST", "EET",  "EEST",
    "MSK",  "MSD",  "AST",  "ADT",  "NST",
    "NDT",  "EST",  "EDT",  "CST",  "CDT",
    "MST",  "MDT",  "PST",  "PDT",  "HST",
    "AKST", "AKDT", "AEST", "AEDT", "ACST",
    "ACDT", "AWST"
};

int inc[32] = {
       0,    0,   60,   60,    0, 
      60,   60,  120,  120,  180, 
     180,  240, -240, -180, -210, 
    -150, -300, -240, -360, -300, 
    -420, -360, -480, -420, -600, 
    -540, -480,  600,  660,  570, 
     630,  480
};

int getmin(string s) {
    int pos = s.find(':');
    int a = atoi(s.substr(0, pos).c_str());
    if(a == 12) a = 0;
    int b = atoi(s.substr(pos + 1).c_str());
    return a * 60 + b;
}

int tomin(string a, string b) {
    if(a == "noon") return 720;
    if(a == "midnight") return 0;
    if(b == "a.m.") return getmin(a);
    else return getmin(a) + 720;
}

void printtime(int a) {
    if(a == 720) puts("noon");
    else if(a == 0) puts("midnight");
    else if(a >= 720) printf("%d:%02d p.m.\n", (a / 60 - 12) ? (a / 60 - 12) : 12, a % 60);
    else printf("%d:%02d a.m.\n", a / 60 ? a / 60 : 12, a % 60);
}

int main() {
    int N;
    for(int i = 0; i < 32; ++i) MP[name[i]] = (inc[i] + 1440) % 1440;
    string a, b, c, d;
    for(cin >> N; N--;) {
        cin >> a;
        if(a != "noon" && a != "midnight") cin >> b;
        cin >> c >> d;
        int t = tomin(a, b);
        t = (1440 + t - MP[c] + MP[d]) % 1440;
        printtime(t);
    }
}

File: 范例/Accpt/19xx/p1916.pas
PROGRAM p1916;

CONST
        name  : array[1..32] of string =
        ('UTC', 'GMT', 'BST', 'IST', 'WET','WEST', 'CET','CEST', 'EET','EEST',
        'MSK' , 'MSD', 'AST', 'ADT', 'NST', 'NDT', 'EST', 'EDT', 'CST', 'CDT',
        'MST' , 'MDT', 'PST', 'PDT', 'HST','AKST','AKDT','AEST','AEDT','ACST',
        'ACDT','AWST');
        change : array[1..32] of Real =
        (    0,     0,     1,      1,     0,    1,     1,     2,     2,     3,
             3,     4,    -4,     -3,  -3.5, -2.5,    -5,    -4,    -6,    -5,
            -7,    -6,    -8,     -7,   -10,   -9,    -8,    10,    11,   9.5,
          10.5,     8);

VAR
        part, p , code , h , m  : Longint;
        s , t                   : String;
        i                       : Integer;

// Make Time well
PROCEDURE Push ( var h , m :Longint ) ;
begin
if m >= 60 then
        begin
        m := m - 60;
        h := h +  1;
        end;
if m < 0 then
        begin
        m := m + 60;
        h := h -  1;
        end;
if h >= 24 then h := h - 24;
if h < 0 then h := h + 24;
end;

PROCEDURE Print ( h , m : Longint ) ;
begin
if ( h = 0 ) and ( m = 0 ) then
        begin
        writeln ( 'midnight' );
        exit;
        end;

if ( h = 12 ) and ( m = 0 ) then
        begin
        writeln ( 'noon' );
        exit;
        end;

if h >= 12 then
        begin
        if h > 12 then h := h - 12;
        write ( h,':');
        if m < 10 then write ( 0 );
        writeln ( m ,' p.m.');
        exit;
        end;

if h = 0 then h := 12;
write ( h ,':' );
if m < 10 then write ( 0 );
writeln ( m ,' a.m.');
end;

// delete the more space
PROCEDURE dowith ( var s :String );
var
        i                       : Longint;
begin
i := 2;
while ( i <= length ( s ) ) do
        if ( s[i-1] = ' ' ) and ( s[i] = ' ' ) then
                delete ( s , i , 1)
                else inc ( i );

while ( s[length(s)] = ' ' ) do
        delete ( s , length (s) , 1 );
end;

PROCEDURE Put_S_To_h_m ;
begin
if pos ('noon' , s ) > 0 then
        begin
        h := 12;
        m := 0;
        Delete ( s , 1 , pos (' ' , s ) );
        exit;
        end;

if pos ('midnight' , s ) > 0 then
        begin
        h := 0;
        m := 0;
        Delete ( s , 1 , pos (' ' , s ) );
        exit;
        end;

val ( copy ( s , 1 , pos ( ':' , s ) - 1 ) , h , code );
Delete ( s , 1 , pos ( ':' , s ) );

val ( copy ( s , 1 , pos ( ' ' , s ) - 1 ) , m , code );
Delete ( s , 1 , pos ( ' ' , s ) );

if pos ( 'p.m.' , s ) > 0 then
        begin
        if h <> 12 then inc ( h , 12 );
        end else
        if h = 12 then h := 0;

Delete ( s , 1 , pos ( ' ' , s ) );
end;

FUNCTION Find ( t : String ) : Integer;
var
        i                       : Integer;
begin
for i:= 1to 32 do if name[i] = t then exit ( i );
end;

PROCEDURE Change_Time;
begin
i := Find (copy ( s , 1 , pos ( ' ' , s ) - 1 ) );
m := round ( m - frac ( change [i] ) * 60 );
h := h - trunc ( change [i] );
push ( h , m );
Delete ( s , 1 , pos ( ' ' , s  ) );

i := Find ( s );
m := round ( m + frac ( change [i] ) * 60 );
h := h + trunc ( change [i] );
push ( h , m );
end;

BEGIN
readln ( part );
for p := 1 to part do
        begin

        readln ( s );

        dowith ( s ); // Make the input very fixful

        Put_S_To_h_m; // h = hour     m = minutes

        Change_Time;  // Do twice one change to UTC one change to the goal

        Print ( h , m );  //  Just output
        end;
END.